lib/libc/newlib/libc-hooks.c

```diff
@@ -157,3 +162,51 @@ void *_sbrk(int count)
 	}
 }
 FUNC_ALIAS(_sbrk, sbrk, void *);
+
+int _gettimeofday(struct timeval *tv, void *tz)
+{
+#if CONFIG_RTC
+	struct device *rtc_dev = NULL;
+
+	rtc_dev = device_get_binding(CONFIG_RTC_0_NAME);
+	if (!rtc_dev) {
+		errno = ENXIO;
+		return -1;
+	}
+
+	if (tv) {
+		tv->tv_sec = rtc_read(rtc_dev);
```

Comment:
This works, but note that on most OSes this function is used for hot path benchmarking and stuff: it tends to be implemented with a layer above the RTC hardware (e.g. on x86 linux it's done with kernel-maintained calibration on top of the TSC and isn't a system call at all).

It's not unlikely that you'll find your users complaining about the slowness of this function more than thanking you for the implementation.

Reply:
@andyross, Thanks. I got the idea. The resolution is not enough. And access RTC hardware directly might not be a good idea because gettimeofday() might be called very frequently.

Reply:
Reading the RTC hardware directly is fine for a first implementation, or if you need this in a pinch.

A better thing to do, IMHO, would Zephyr count the time itself to a 64-bit integer, and have these functions query those numbers.  This would work even on devices whose RTC registers a 32-bit wide.  Also, with the memory protection and system call thing becoming reality, the lesser we have direct hardware access for high level concepts such as time, the better.